app/scripts/metamask-controller.js

```diff
@@ -5498,6 +5567,265 @@ export default class MetamaskController extends EventEmitter {
     return engine;
   }
 
+  /**
+   * A method for creating a provider that is safely restricted for the requesting subject.
+   *
+   * @param {object} options - Provider engine options
+   * @param {string} options.origin - The origin of the sender
+   * @param {tabId} [options.tabId] - The tab ID of the sender - if the sender is within a tab
+   */
+  setupProviderEngineCaip({ origin, tabId }) {
+    const engine = new JsonRpcEngine();
```

Comment:
Can we DRY a bunch of what's common between these 2 provider engines? Or I guess that might defeat the purpose of separating them... 🤔

Reply:
yeah it's hard to say. They have a very similar shape. Some of it can probably be DRYed but i'm hoping to deal with that later as this isn't quite ready for review yet